File: 189-rotate-array/rotate-array.cpp
class Solution {
public:
    void rotate(vector<int>& nums, int k) {
// The Brute Force 
        // int n = nums.size();
        // if(n == 0) return;
        // k = k % n;
        // if (k > n)
        //     return;
        // int temp[k]; // required temp array of size k 
        // // we start shifting from kth index
        // for(int i = k; i < n; i++){
        //     nums[i - k] = nums[i];
        // }

        // // puting back temp array

        // for (int i = n - k; i < n; i++){
        //     nums[i] = temp[i - (n - k)];
        // }

        // Optimal solution : 

        k %= nums.size();
        reverse(nums.begin(), nums.end());
        reverse(nums.begin(), nums.begin() + k);
        reverse(nums.begin() + k, nums.end());
    }
        

};

File: 189-rotate-array/rotate-array.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var rotate = function(nums, k) {
    // Brute force with O(N) TC and SC
    // let n = nums.length;
    // let temp = [];
    // for (let i = 0; i < n; i++)
    // {
    //     temp[(i+k) % n] = nums[i];
    // }
    // for( let j = 0; j < n; j++)
    // {
    //     nums[j] = temp[j];
    // }
    // return nums;

    // optimal one 
    let n = nums.length;
    k = k % n;
    
    nums.reverse();
    reversefun(nums, 0, k - 1);
    reversefun(nums, k , n - 1);

    function reversefun (nums, l, r )
    {
        while ( l < r)
        {
            [nums [l], nums[r]] = [nums[r], nums[l]];
            l++;
            r--;
        }
    } 
    return nums;

};